README.md
ember-async-observer-polyfill
==============================================================================

Implement's [emberjs/rfcs#494](https://github.com/emberjs/rfcs/blob/master/text/0494-async-observers.md) for older Ember versions.


Compatibility
------------------------------------------------------------------------------

* Completely inert when running Ember 3.13 or higher
* Supports Ember.js 2.12, 2.16, 2.18, 3.4, 3.8, 3.12
* Ember CLI v2.13 or above
* Node.js v10 or above


Installation
------------------------------------------------------------------------------

```
ember install ember-async-observer-polyfill
```


Usage
------------------------------------------------------------------------------

The best usage guide is the RFC itself [emberjs/rfcs#494](https://github.com/emberjs/rfcs/blob/master/text/0494-async-observers.md)
but here are a few examples of "before"/"after" to whet your appetite:

**Before**:

```js
const Person = EmberObject.extend({
  firstName: null,
  lastName: null,

  fullName: computed('firstName', 'lastName', function() {
    return `${this.firstName} ${this.lastName}`;
  }),


  fullNameChanged: observer('fullName', function() {
    // do something :)
  }),
});
```

**After**:

```js
const Person = EmberObject.extend({
  firstName: null,
  lastName: null,

  fullName: computed('firstName', 'lastName', function() {
    return `${this.firstName} ${this.lastName}`;
  }),


  fullNameChanged: observer({
    sync: false,
    dependentKeys: ['fullName'],
    fn() {
      // do something :)
    }
  }),
});
```

Limitations
------------------------------------------------------------------------------

There are no known limitations, all features described in the RFC are polyfilled.

Contributing
------------------------------------------------------------------------------

See the [Contributing](CONTRIBUTING.md) guide for details.


License
------------------------------------------------------------------------------

This project is licensed under the [MIT License](LICENSE.md).


CHANGELOG.md
## v1.0.0 (2020-01-03)

#### :boom: Breaking Change
* [#3](https://github.com/ember-polyfills/ember-async-observers-polyfill/pull/3) Drop Node 8 support. ([@rwjblue](https://github.com/rwjblue))

#### :rocket: Enhancement
* [#1](https://github.com/ember-polyfills/ember-async-observers-polyfill/pull/1) Add polyfill implementation for `Ember.observes` ([@rwjblue](https://github.com/rwjblue))

#### :memo: Documentation
* [#2](https://github.com/ember-polyfills/ember-async-observers-polyfill/pull/2) Flesh out basic information in README.md. ([@rwjblue](https://github.com/rwjblue))

#### Committers: 1
- Robert Jackson ([@rwjblue](https://github.com/rwjblue))



tests/test-helper.js
import Application from '../app';
import config from '../config/environment';
import { setApplication } from '@ember/test-helpers';
import { start } from 'ember-qunit';

setApplication(Application.create(config.APP));

start({
  setupTestIsolationValidation: true
});


tests/unit/async-observer-test.js
/* eslint-disable ember/no-observers */

import { settled } from '@ember/test-helpers';
import EmberObject, { computed, observer } from '@ember/object';
import QUnit, { module, test } from 'qunit';

module('ember-async-observer-polyfill', function() {
  module('observer', function() {
    let buildObserver;

    function buildPerson(options) {
      let Person = EmberObject.extend({
        firstName: null,
        lastName: null,

        fullName: computed('firstName', 'lastName', function() {
          return `${this.firstName} ${this.lastName}`;
        }),

        fullNameChanged: buildObserver(),
      });

      return Person.create(options);
    }

    test('runs sync with legacy API', function(assert) {
      buildObserver = () =>
        observer('fullName', function() {
          assert.step(`fired for: ${this.get('fullName')}`);
        });

      let person = buildPerson({ firstName: 'max', lastName: 'jackson' });
      person.get('fullName'); // consume observed property

      assert.verifySteps([], 'precond - observer not fired');
      person.set('firstName', 'james');
      assert.verifySteps(['fired for: james jackson'], 'observer after upstream property updated');

      assert.strictEqual(person.get('fullName'), 'james jackson');
    });

    test('errors without sync specified in emberjs/rfcs#494 API', function(assert) {
      assert.throws(() => {
        observer({
          dependentKeys: ['fullName'],
          fn() {
            QUnit.config.current.assert.step(`fired for: ${this.fullName}`);
          },
        });
      }, /observer called without sync/);
    });

    test('runs sync with emberjs/rfcs#494 API', function(assert) {
      buildObserver = () =>
        observer({
          sync: true,
          dependentKeys: ['fullName'],
          fn() {
            assert.step(`fired for: ${this.get('fullName')}`);
          },
        });

      let person = buildPerson({ firstName: 'max', lastName: 'jackson' });
      person.get('fullName'); // consume observed property

      assert.verifySteps([], 'precond - observer not fired');
      person.set('firstName', 'james');
      assert.verifySteps(['fired for: james jackson'], 'observer after upstream property updated');

      assert.strictEqual(person.get('fullName'), 'james jackson');
    });

    test('runs async with emberjs/rfcs#494 API', async function(assert) {
      buildObserver = () =>
        observer({
          sync: false,
          dependentKeys: ['fullName'],
          fn(...args) {
            assert.deepEqual(args, [this, 'fullName'], 'arguments are correct');
            assert.step(`fired for: ${this.get('fullName')}`);
          },
        });

      let person = buildPerson({ firstName: 'max', lastName: 'jackson' });
      person.get('fullName'); // consume observed property

      assert.verifySteps([], 'precond - observer not fired');
      person.set('firstName', 'james');
      assert.verifySteps([], 'precond - observer not fired');

      await settled();

      assert.verifySteps(['fired for: james jackson'], 'observer after upstream property updated');

      assert.strictEqual(person.get('fullName'), 'james jackson');
    });
  });
});


index.js
'use strict';

const VersionChecker = require('ember-cli-version-checker');

module.exports = {
  name: require('./package').name,

  init() {
    this._super.init && this._super.init.apply(this, arguments);

    let checker = new VersionChecker(this.project);
    let emberVersion = checker.forEmber();

    this.shouldPolyfill = emberVersion.lt('3.13.0-alpha.1');
  },

  included() {
    this._super.included.apply(this, arguments);

    if (!this.shouldPolyfill) {
      return;
    }

    // grab the original _findHost, before ember-engines can muck with it
    let proto = Object.getPrototypeOf(this);
    let host = proto._findHost.call(this);

    // using `app.import` here because we need to ensure this ends up in the
    // top level vendor.js (not engine-vendor.js as would happen with
    // this.import inside an engine)
    host.import('vendor/ember-async-observer-polyfill/index.js');
  },

  treeForVendor(rawVendorTree) {
    if (!this.shouldPolyfill) {
      return;
    }

    let babelAddon = this.addons.find(addon => addon.name === 'ember-cli-babel');

    let transpiledVendorTree = babelAddon.transpileTree(rawVendorTree, {
      babel: this.options.babel,

      'ember-cli-babel': {
        compileModules: false,
      },
    });

    return transpiledVendorTree;
  },
};


.eslintrc.js
module.exports = {
  root: true,
  parser: 'babel-eslint',
  parserOptions: {
    ecmaVersion: 2018,
    sourceType: 'module',
    ecmaFeatures: {
      legacyDecorators: true,
    },
  },
  plugins: ['ember', 'prettier'],
  extends: ['eslint:recommended', 'plugin:ember/recommended', 'plugin:prettier/recommended'],
  env: {
    browser: true,
  },
  rules: {
    'ember/no-jquery': 'error',
  },
  overrides: [
    // node files
    {
      files: [
        '.eslintrc.js',
        '.template-lintrc.js',
        'ember-cli-build.js',
        'index.js',
        'testem.js',
        'blueprints/*/index.js',
        'config/**/*.js',
        'tests/dummy/config/**/*.js',
      ],
      excludedFiles: [
        'addon/**',
        'addon-test-support/**',
        'app/**',
        'tests/dummy/app/**',
        'vendor/**',
      ],
      parserOptions: {
        sourceType: 'script',
      },
      env: {
        browser: false,
        node: true,
      },
      plugins: ['node'],
      rules: Object.assign({}, require('eslint-plugin-node').configs.recommended.rules, {
        // add your custom rules and overrides for node files here
      }),
    },
  ],
};


config/ember-try.js
'use strict';

const getChannelURL = require('ember-source-channel-url');

module.exports = async function() {
  return {
    useYarn: true,
    scenarios: [
      {
        name: 'ember-lts-2.12',
        env: {
          EMBER_OPTIONAL_FEATURES: JSON.stringify({ 'jquery-integration': true }),
        },
        npm: {
          devDependencies: {
            '@ember/jquery': '^0.5.1',
            'ember-source': '~2.12.0',
          },
        },
      },
      {
        name: 'ember-lts-2.16',
        env: {
          EMBER_OPTIONAL_FEATURES: JSON.stringify({ 'jquery-integration': true }),
        },
        npm: {
          devDependencies: {
            '@ember/jquery': '^0.5.1',
            'ember-source': '~2.16.0',
          },
        },
      },
      {
        name: 'ember-lts-2.18',
        env: {
          EMBER_OPTIONAL_FEATURES: JSON.stringify({ 'jquery-integration': true }),
        },
        npm: {
          devDependencies: {
            '@ember/jquery': '^0.5.1',
            'ember-source': '~2.18.0',
          },
        },
      },
      {
        name: 'ember-lts-3.4',
        npm: {
          devDependencies: {
            'ember-source': '~3.8.0',
          },
        },
      },
      {
        name: 'ember-lts-3.8',
        npm: {
          devDependencies: {
            'ember-source': '~3.8.0',
          },
        },
      },
      {
        name: 'ember-lts-3.12',
        npm: {
          devDependencies: {
            'ember-source': '~3.12.0',
          },
        },
      },
      {
        name: 'ember-release',
        npm: {
          devDependencies: {
            'ember-source': await getChannelURL('release'),
          },
        },
      },
      {
        name: 'ember-beta',
        npm: {
          devDependencies: {
            'ember-source': await getChannelURL('beta'),
          },
        },
      },
      {
        name: 'ember-canary',
        npm: {
          devDependencies: {
            'ember-source': await getChannelURL('canary'),
          },
        },
      },
      // The default `.travis.yml` runs this scenario via `yarn test`,
      // not via `ember try`. It's still included here so that running
      // `ember try:each` manually or from a customized CI config will run it
      // along with all the other scenarios.
      {
        name: 'ember-default',
        npm: {
          devDependencies: {},
        },
      },
      {
        name: 'ember-default-with-jquery',
        env: {
          EMBER_OPTIONAL_FEATURES: JSON.stringify({
            'jquery-integration': true,
          }),
        },
        npm: {
          devDependencies: {
            '@ember/jquery': '^0.5.1',
          },
        },
      },
      {
        name: 'ember-classic',
        env: {
          EMBER_OPTIONAL_FEATURES: JSON.stringify({
            'application-template-wrapper': true,
            'default-async-observers': false,
            'template-only-glimmer-components': false,
          }),
        },
        npm: {
          ember: {
            edition: 'classic',
          },
        },
      },
    ],
  };
};
